File: consumer.js
// consumer.js — Polls SQS and processes orders one by one
// Run: node consumer.js

const {
  SQSClient,
  ReceiveMessageCommand,
  DeleteMessageCommand,
} = require("@aws-sdk/client-sqs");

const client = new SQSClient({ region: "us-east-1" });

const QUEUE_URL = process.env.SQS_QUEUE_URL;

// Simulates DB write / payment processing
// Randomly fails 10% of the time to demo DLQ behavior
async function processOrder(order) {
  console.log(`\n⚙️  Processing order: ${order.orderId}`);
  console.log(`   Product  : ${order.product} x${order.quantity}`);
  console.log(`   User     : ${order.userId}`);

  // Simulate processing time (0.5s – 1.5s)
  await new Promise((res) => setTimeout(res, 500 + Math.random() * 1000));

  // Simulate 10% failure rate — increase to 0.5 to demo DLQ
  if (Math.random() < 0.1) {
    throw new Error(`Payment gateway timeout for order ${order.orderId}`);
  }

  console.log(`✅ Order ${order.orderId} fulfilled!`);
}

async function pollQueue() {
  console.log("🔄 Consumer started. Polling SQS queue...\n");

  while (true) {
    const response = await client.send(
      new ReceiveMessageCommand({
        QueueUrl: QUEUE_URL,
        MaxNumberOfMessages: 5,    // batch up to 5 at a time
        WaitTimeSeconds: 20,       // long polling — saves API cost
        VisibilityTimeout: 30,     // hide msg for 30s while processing
      })
    );

    const messages = response.Messages || [];

    if (messages.length === 0) {
      console.log("📭 Queue empty. Waiting for orders...");
      continue;
    }

    console.log(`📬 Received ${messages.length} message(s)`);

    for (const message of messages) {
      const order = JSON.parse(message.Body);

      try {
        await processOrder(order);

        // Delete ONLY after successful processing
        await client.send(
          new DeleteMessageCommand({
            QueueUrl: QUEUE_URL,
            ReceiptHandle: message.ReceiptHandle,
          })
        );

        console.log(`🗑️  Message deleted from queue`);
      } catch (err) {
        // DO NOT delete — SQS will retry after VisibilityTimeout expires
        // After maxReceiveCount (3) retries, it moves to DLQ automatically
        console.error(`❌ Failed: ${err.message}`);
        console.error(`   Order will be retried by SQS...`);
      }
    }
  }
}

pollQueue().catch(console.error);

File: consumer.py
# consumer.py — Polls SQS and processes orders one by one
# Run: python consumer.py
# Install: pip install boto3

import boto3
import json
import time
import random
import os

sqs = boto3.client("sqs", region_name="us-east-1")

QUEUE_URL = os.environ["SQS_QUEUE_URL"]


def process_order(order):
    """Simulates DB write + payment. Fails 10% of the time for DLQ demo."""
    print(f"\n⚙️  Processing order: {order['orderId']}")
    print(f"   Product  : {order['product']} x{order['quantity']}")
    print(f"   User     : {order['userId']}")

    time.sleep(random.uniform(0.5, 1.5))

    # Simulate 10% failure rate — increase to 0.5 to demo DLQ
    if random.random() < 0.1:
        raise Exception(f"Payment gateway timeout for order {order['orderId']}")

    print(f"✅ Order {order['orderId']} fulfilled!")


def poll_queue():
    print("🔄 Consumer started. Polling SQS queue...\n")

    while True:
        response = sqs.receive_message(
            QueueUrl=QUEUE_URL,
            MaxNumberOfMessages=5,
            WaitTimeSeconds=20,
            VisibilityTimeout=30,
        )

        messages = response.get("Messages", [])

        if not messages:
            print("📭 Queue empty. Waiting for orders...")
            continue

        print(f"📬 Received {len(messages)} message(s)")

        for message in messages:
            order = json.loads(message["Body"])

            try:
                process_order(order)

                sqs.delete_message(
                    QueueUrl=QUEUE_URL,
                    ReceiptHandle=message["ReceiptHandle"],
                )
                print("🗑️  Message deleted from queue")

            except Exception as e:
                # DO NOT delete — SQS retries after VisibilityTimeout
                # After maxReceiveCount (3) retries → moves to DLQ
                print(f"❌ Failed: {e}")
                print("   Order will be retried by SQS...")


if __name__ == "__main__":
    poll_queue()
